### PR TITLE
Feat/biconomy modular smart account

### PR DESCRIPTION
**Pull request by livingrockrises**
Add methods to use Biconomy Smart Account which uses ECDSA module as default validation module.

### CODE REVIEW

**Comment by livingrockrises** [comment]
I have fixed some linting issues in utils and in one of the kernel files.

**Comment by livingrockrises** [comment]
tests
Screenshot 2023-12-07 at 7 25 35 PM

File: src/accounts/biconomy/abi/BiconomySmartAccountAbi.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                name: "value",
+                type: "uint256[]"
+            },
+            {
+                internalType: "bytes[]",
+                name: "func",
+                type: "bytes[]"
+            }
+        ],
+        name: "executeBatch",
```
Comment (plusminushalf) [inline src/accounts/biconomy/abi/BiconomySmartAccountAbi.ts:69]: I think we don't use this function of the abi, we can remove this.

File: src/accounts/biconomy/signerToBiconomySmartAccount.ts [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+    {
+        signer,
+        entryPoint,
+        index = 0n,
+        factoryAddress = BICONOMY_ADDRESSES.FACTORY_ADDRESS,
+        accountLogicAddress = BICONOMY_ADDRESSES.ACCOUNT_V2_0_LOGIC,
+        fallbackHandlerAddress = BICONOMY_ADDRESSES.DEFAULT_FALLBACK_HANDLER_ADDRESS,
+        ecdsaModuleAddress = BICONOMY_ADDRESSES.ECDSA_OWNERSHIP_REGISTRY_MODULE
+    }: {
+        signer: SmartAccountSigner
```
Comment (plusminushalf) [inline src/accounts/biconomy/signerToBiconomySmartAccount.ts:235]: `SmartAccountSigner` now needs source & address;

File: src/accounts/biconomy/signerToBiconomySmartAccount.ts [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+    client: Client<TTransport, TChain>
+    factoryAddress: Address
+    accountLogicAddress: Address
+    fallbackHandlerAddress: Address
+    ecdsaModuleAddress: Address
+    owner: Address
+    entryPoint: Address
+    index?: bigint
+}): Promise<Address> => {
+    // @ts-ignore
```
Comment (plusminushalf) [inline src/accounts/biconomy/signerToBiconomySmartAccount.ts:159]: need to resolve this.

File: src/accounts/biconomy/signerToBiconomySmartAccount.ts [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+    const unusedClient = client // Ignoring the error for now
+    // @ts-ignore
```
Comment (plusminushalf) [inline src/accounts/biconomy/signerToBiconomySmartAccount.ts:161]: this too

File: src/accounts/biconomy/signerToBiconomySmartAccount.ts [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+        factoryAddress,
+        encodeFunctionData({
+            abi: createAccountAbi,
+            functionName: "deployCounterFactualAccount",
+            args: [ecdsaModuleAddress, ecdsaOwnershipInitData, index]
+        }) as Hex
+    ])
+}
+
+const getAccountAddress = async <
```
Comment (plusminushalf) [inline src/accounts/biconomy/signerToBiconomySmartAccount.ts:137]: Can we not use:

```typescript
const initCode = await getAccountInitCode(...)

return getSenderAddress(client, {
    initCode,
    entryPoint
})
```

**Review by plusminushalf — CHANGES_REQUESTED** [review]
Also we can remove commented code.

File: src/accounts/biconomy/abi/BiconomySmartAccountAbi.ts (hunk 1 above)
Comment (livingrockrises) [reply src/accounts/biconomy/abi/BiconomySmartAccountAbi.ts:69]: ok will keep only optimised selectors.

File: src/accounts/biconomy/signerToBiconomySmartAccount.ts (hunk 2 above)
Comment (livingrockrises) [reply src/accounts/biconomy/signerToBiconomySmartAccount.ts:235]: can you share an example or link particular line/s in latest code

File: src/accounts/biconomy/signerToBiconomySmartAccount.ts (hunk 3 above)
Comment (livingrockrises) [reply src/accounts/biconomy/signerToBiconomySmartAccount.ts:159]: currently I don't see use of entryPoint or client but we may require client. hence I have kept it to be consistent with rest of interfaces

File: src/accounts/biconomy/signerToBiconomySmartAccount.ts (hunk 5 above)
Comment (livingrockrises) [reply src/accounts/biconomy/signerToBiconomySmartAccount.ts:137]: I actually feel current way is more efficient and doesn't require any network calls.

File: src/accounts/biconomy/signerToBiconomySmartAccount.ts (hunk 2 above)
Comment (livingrockrises) [reply src/accounts/biconomy/signerToBiconomySmartAccount.ts:235]: ok I see

Comment (livingrockrises) [reply src/accounts/biconomy/signerToBiconomySmartAccount.ts:235]: fixed

File: src/accounts/biconomy/abi/BiconomySmartAccountAbi.ts (hunk 1 above)
Comment (livingrockrises) [reply src/accounts/biconomy/abi/BiconomySmartAccountAbi.ts]: done

File: src/accounts/biconomy/signerToBiconomySmartAccount.ts (hunk 5 above)
Comment (livingrockrises) [reply src/accounts/biconomy/signerToBiconomySmartAccount.ts:137]: also there should not be a need to pass deployedAccountAddress the way it is done in some other helpers and tests.

File: src/accounts/biconomy/signerToBiconomySmartAccount.ts (hunk 4 above)
Comment (livingrockrises) [reply src/accounts/biconomy/signerToBiconomySmartAccount.ts:161]: fixed

File: src/accounts/biconomy/signerToBiconomySmartAccount.ts (hunk 3 above)
Comment (livingrockrises) [reply src/accounts/biconomy/signerToBiconomySmartAccount.ts:159]: for now I have removed these and fixed.

**Review by plusminushalf — APPROVED** [review]

**Pull request merged**